File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-fx-5-4/ExchangeRate.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class ExchangeRate : ISerialized
    {
        public ExchangeRate(XmlNode xmlNode)
        {
            XmlNode quotedCurrencyPairNode = xmlNode.SelectSingleNode("quotedCurrencyPair");
            
            if (quotedCurrencyPairNode != null)
            {
                if (quotedCurrencyPairNode.Attributes["href"] != null || quotedCurrencyPairNode.Attributes["id"] != null) 
                {
                    if (quotedCurrencyPairNode.Attributes["id"] != null) 
                    {
                        quotedCurrencyPairIDRef_ = quotedCurrencyPairNode.Attributes["id"].Value;
                        QuotedCurrencyPair ob = new QuotedCurrencyPair(quotedCurrencyPairNode);
                        IDManager.SetID(quotedCurrencyPairIDRef_, ob);
                    }
                    else if (quotedCurrencyPairNode.Attributes["href"] != null)
                    {
                        quotedCurrencyPairIDRef_ = quotedCurrencyPairNode.Attributes["href"].Value;
                    }
                    else
                    {
                        quotedCurrencyPair_ = new QuotedCurrencyPair(quotedCurrencyPairNode);
                    }
                }
                else
                {
                    quotedCurrencyPair_ = new QuotedCurrencyPair(quotedCurrencyPairNode);
                }
            }
            
        
            XmlNode rateNode = xmlNode.SelectSingleNode("rate");
            
            if (rateNode != null)
            {
                if (rateNode.Attributes["href"] != null || rateNode.Attributes["id"] != null) 
                {
                    if (rateNode.Attributes["id"] != null) 
                    {
                        rateIDRef_ = rateNode.Attributes["id"].Value;
                        PositiveDecimal ob = new PositiveDecimal(rateNode);
                        IDManager.SetID(rateIDRef_, ob);
                    }
                    else if (rateNode.Attributes["href"] != null)
                    {
                        rateIDRef_ = rateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        rate_ = new PositiveDecimal(rateNode);
                    }
                }
                else
                {
                    rate_ = new PositiveDecimal(rateNode);
                }
            }
            
        
            XmlNode spotRateNode = xmlNode.SelectSingleNode("spotRate");
            
            if (spotRateNode != null)
            {
                if (spotRateNode.Attributes["href"] != null || spotRateNode.Attributes["id"] != null) 
                {
                    if (spotRateNode.Attributes["id"] != null) 
                    {
                        spotRateIDRef_ = spotRateNode.Attributes["id"].Value;
                        PositiveDecimal ob = new PositiveDecimal(spotRateNode);
                        IDManager.SetID(spotRateIDRef_, ob);
                    }
                    else if (spotRateNode.Attributes["href"] != null)
                    {
                        spotRateIDRef_ = spotRateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        spotRate_ = new PositiveDecimal(spotRateNode);
                    }
                }
                else
                {
                    spotRate_ = new PositiveDecimal(spotRateNode);
                }
            }
            
        
            XmlNode forwardPointsNode = xmlNode.SelectSingleNode("forwardPoints");
            
            if (forwardPointsNode != null)
            {
                if (forwardPointsNode.Attributes["href"] != null || forwardPointsNode.Attributes["id"] != null) 
                {
                    if (forwardPointsNode.Attributes["id"] != null) 
                    {
                        forwardPointsIDRef_ = forwardPointsNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(forwardPointsNode);
                        IDManager.SetID(forwardPointsIDRef_, ob);
                    }
                    else if (forwardPointsNode.Attributes["href"] != null)
                    {
                        forwardPointsIDRef_ = forwardPointsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        forwardPoints_ = new XsdTypeDecimal(forwardPointsNode);
                    }
                }
                else
                {
                    forwardPoints_ = new XsdTypeDecimal(forwardPointsNode);
                }
            }
            
        
            XmlNode pointValueNode = xmlNode.SelectSingleNode("pointValue");
            
            if (pointValueNode != null)
            {
                if (pointValueNode.Attributes["href"] != null || pointValueNode.Attributes["id"] != null) 
                {
                    if (pointValueNode.Attributes["id"] != null) 
                    {
                        pointValueIDRef_ = pointValueNode.Attributes["id"].Value;
                        PointValue ob = new PointValue(pointValueNode);
                        IDManager.SetID(pointValueIDRef_, ob);
                    }
                    else if (pointValueNode.Attributes["href"] != null)
                    {
                        pointValueIDRef_ = pointValueNode.Attributes["href"].Value;
                    }
                    else
                    {
                        pointValue_ = new PointValue(pointValueNode);
                    }
                }
                else
                {
                    pointValue_ = new PointValue(pointValueNode);
                }
            }
            
        
            XmlNodeList crossRateNodeList = xmlNode.SelectNodes("crossRate");
            
            if (crossRateNodeList != null)
            {
                this.crossRate_ = new List<CrossRate>();
                foreach (XmlNode item in crossRateNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            crossRateIDRef_ = item.Attributes["id"].Value;
                            crossRate_.Add(new CrossRate(item));
                            IDManager.SetID(crossRateIDRef_, crossRate_[crossRate_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            crossRateIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        crossRate_.Add(new CrossRate(item));
                        }
                    }
                    else
                    {
                        crossRate_.Add(new CrossRate(item));
                    }
                }
            }
            
        
        }
        
    
        #region QuotedCurrencyPair_
        private QuotedCurrencyPair quotedCurrencyPair_;
        public QuotedCurrencyPair QuotedCurrencyPair_
        {
            get
            {
                if (this.quotedCurrencyPair_ != null)
                {
                    return this.quotedCurrencyPair_; 
                }
                else if (this.quotedCurrencyPairIDRef_ != null)
                {
                    quotedCurrencyPair_ = IDManager.getID(quotedCurrencyPairIDRef_) as QuotedCurrencyPair;
                    return this.quotedCurrencyPair_; 
                }
                else
                {
                      return this.quotedCurrencyPair_; 
                }
            }
            set
            {
                if (this.quotedCurrencyPair_ != value)
                {
                    this.quotedCurrencyPair_ = value;
                }
            }
        }
        #endregion
        
        public string quotedCurrencyPairIDRef_ { get; set; }
        #region Rate_
        private PositiveDecimal rate_;
        public PositiveDecimal Rate_
        {
            get
            {
                if (this.rate_ != null)
                {
                    return this.rate_; 
                }
                else if (this.rateIDRef_ != null)
                {
                    rate_ = IDManager.getID(rateIDRef_) as PositiveDecimal;
                    return this.rate_; 
                }
                else
                {
                      return this.rate_; 
                }
            }
            set
            {
                if (this.rate_ != value)
                {
                    this.rate_ = value;
                }
            }
        }
        #endregion
        
        public string rateIDRef_ { get; set; }
        #region SpotRate_
        private PositiveDecimal spotRate_;
        public PositiveDecimal SpotRate_
        {
            get
            {
                if (this.spotRate_ != null)
                {
                    return this.spotRate_; 
                }
                else if (this.spotRateIDRef_ != null)
                {
                    spotRate_ = IDManager.getID(spotRateIDRef_) as PositiveDecimal;
                    return this.spotRate_; 
                }
                else
                {
                      return this.spotRate_; 
                }
            }
            set
            {
                if (this.spotRate_ != value)
                {
                    this.spotRate_ = value;
                }
            }
        }
        #endregion
        
        public string spotRateIDRef_ { get; set; }
        #region ForwardPoints_
        private XsdTypeDecimal forwardPoints_;
        public XsdTypeDecimal ForwardPoints_
        {
            get
            {
                if (this.forwardPoints_ != null)
                {
                    return this.forwardPoints_; 
                }
                else if (this.forwardPointsIDRef_ != null)
                {
                    forwardPoints_ = IDManager.getID(forwardPointsIDRef_) as XsdTypeDecimal;
                    return this.forwardPoints_; 
                }
                else
                {
                      return this.forwardPoints_; 
                }
            }
            set
            {
                if (this.forwardPoints_ != value)
                {
                    this.forwardPoints_ = value;
                }
            }
        }
        #endregion
        
        public string forwardPointsIDRef_ { get; set; }
        #region PointValue_
        private PointValue pointValue_;
        public PointValue PointValue_
        {
            get
            {
                if (this.pointValue_ != null)
                {
                    return this.pointValue_; 
                }
                else if (this.pointValueIDRef_ != null)
                {
                    pointValue_ = IDManager.getID(pointValueIDRef_) as PointValue;
                    return this.pointValue_; 
                }
                else
                {
                      return this.pointValue_; 
                }
            }
            set
            {
                if (this.pointValue_ != value)
                {
                    this.pointValue_ = value;
                }
            }
        }
        #endregion
        
        public string pointValueIDRef_ { get; set; }
        #region CrossRate_
        private List<CrossRate> crossRate_;
        public List<CrossRate> CrossRate_
        {
            get
            {
                if (this.crossRate_ != null)
                {
                    return this.crossRate_; 
                }
                else if (this.crossRateIDRef_ != null)
                {
                    return this.crossRate_; 
                }
                else
                {
                      return this.crossRate_; 
                }
            }
            set
            {
                if (this.crossRate_ != value)
                {
                    this.crossRate_ = value;
                }
            }
        }
        #endregion
        
        public string crossRateIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-fx-5-4/ExchangeRate.cpp
// ExchangeRate.cpp 
#include "ExchangeRate.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

ExchangeRate::ExchangeRate(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //quotedCurrencyPairNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* quotedCurrencyPairNode = xmlNode->FirstChildElement("quotedCurrencyPair");

   if(quotedCurrencyPairNode){quotedCurrencyPairIsNull_ = false;}
   else{quotedCurrencyPairIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- quotedCurrencyPairNode , address : " << quotedCurrencyPairNode << std::endl;
   #endif
   if(quotedCurrencyPairNode)
   {
      if(quotedCurrencyPairNode->Attribute("href") || quotedCurrencyPairNode->Attribute("id"))
      {
          if(quotedCurrencyPairNode->Attribute("id"))
          {
             quotedCurrencyPairIDRef_ = quotedCurrencyPairNode->Attribute("id");
             quotedCurrencyPair_ = boost::shared_ptr<QuotedCurrencyPair>(new QuotedCurrencyPair(quotedCurrencyPairNode));
             quotedCurrencyPair_->setID(quotedCurrencyPairIDRef_);
             IDManager::instance().setID(quotedCurrencyPairIDRef_,quotedCurrencyPair_);
          }
          else if(quotedCurrencyPairNode->Attribute("href")) { quotedCurrencyPairIDRef_ = quotedCurrencyPairNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { quotedCurrencyPair_ = boost::shared_ptr<QuotedCurrencyPair>(new QuotedCurrencyPair(quotedCurrencyPairNode));}
   }

   //rateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* rateNode = xmlNode->FirstChildElement("rate");

   if(rateNode){rateIsNull_ = false;}
   else{rateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- rateNode , address : " << rateNode << std::endl;
   #endif
   if(rateNode)
   {
      if(rateNode->Attribute("href") || rateNode->Attribute("id"))
      {
          if(rateNode->Attribute("id"))
          {
             rateIDRef_ = rateNode->Attribute("id");
             rate_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(rateNode));
             rate_->setID(rateIDRef_);
             IDManager::instance().setID(rateIDRef_,rate_);
          }
          else if(rateNode->Attribute("href")) { rateIDRef_ = rateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { rate_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(rateNode));}
   }

   //spotRateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* spotRateNode = xmlNode->FirstChildElement("spotRate");

   if(spotRateNode){spotRateIsNull_ = false;}
   else{spotRateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- spotRateNode , address : " << spotRateNode << std::endl;
   #endif
   if(spotRateNode)
   {
      if(spotRateNode->Attribute("href") || spotRateNode->Attribute("id"))
      {
          if(spotRateNode->Attribute("id"))
          {
             spotRateIDRef_ = spotRateNode->Attribute("id");
             spotRate_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(spotRateNode));
             spotRate_->setID(spotRateIDRef_);
             IDManager::instance().setID(spotRateIDRef_,spotRate_);
          }
          else if(spotRateNode->Attribute("href")) { spotRateIDRef_ = spotRateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { spotRate_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(spotRateNode));}
   }

   //forwardPointsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* forwardPointsNode = xmlNode->FirstChildElement("forwardPoints");

   if(forwardPointsNode){forwardPointsIsNull_ = false;}
   else{forwardPointsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- forwardPointsNode , address : " << forwardPointsNode << std::endl;
   #endif
   if(forwardPointsNode)
   {
      if(forwardPointsNode->Attribute("href") || forwardPointsNode->Attribute("id"))
      {
          if(forwardPointsNode->Attribute("id"))
          {
             forwardPointsIDRef_ = forwardPointsNode->Attribute("id");
             forwardPoints_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(forwardPointsNode));
             forwardPoints_->setID(forwardPointsIDRef_);
             IDManager::instance().setID(forwardPointsIDRef_,forwardPoints_);
          }
          else if(forwardPointsNode->Attribute("href")) { forwardPointsIDRef_ = forwardPointsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { forwardPoints_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(forwardPointsNode));}
   }

   //pointValueNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* pointValueNode = xmlNode->FirstChildElement("pointValue");

   if(pointValueNode){pointValueIsNull_ = false;}
   else{pointValueIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- pointValueNode , address : " << pointValueNode << std::endl;
   #endif
   if(pointValueNode)
   {
      if(pointValueNode->Attribute("href") || pointValueNode->Attribute("id"))
      {
          if(pointValueNode->Attribute("id"))
          {
             pointValueIDRef_ = pointValueNode->Attribute("id");
             pointValue_ = boost::shared_ptr<PointValue>(new PointValue(pointValueNode));
             pointValue_->setID(pointValueIDRef_);
             IDManager::instance().setID(pointValueIDRef_,pointValue_);
          }
          else if(pointValueNode->Attribute("href")) { pointValueIDRef_ = pointValueNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { pointValue_ = boost::shared_ptr<PointValue>(new PointValue(pointValueNode));}
   }

   //crossRateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* crossRateNode = xmlNode->FirstChildElement("crossRate");

   if(crossRateNode){crossRateIsNull_ = false;}
   else{crossRateIsNull_ = true;}

   if(crossRateNode)
   {
      for(crossRateNode; crossRateNode; crossRateNode = crossRateNode->NextSiblingElement("crossRate")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- crossRateNode , address : " << crossRateNode << std::endl;
          #endif
          if(crossRateNode->Attribute("href") || crossRateNode->Attribute("id"))
          {
              if(crossRateNode->Attribute("id"))
              {
                  crossRateIDRef_ = crossRateNode->Attribute("id");
                  crossRate_.push_back(boost::shared_ptr<CrossRate>(new CrossRate(crossRateNode)));
                  crossRate_.back()->setID(crossRateIDRef_);
                  IDManager::instance().setID(crossRateIDRef_, crossRate_.back());
              }
              else if(crossRateNode->Attribute("href")) { crossRateIDRef_ = crossRateNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { crossRate_.push_back(boost::shared_ptr<CrossRate>(new CrossRate(crossRateNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- crossRateNode , address : " << crossRateNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<QuotedCurrencyPair> ExchangeRate::getQuotedCurrencyPair()
{
   if(!this->quotedCurrencyPairIsNull_){
        if(quotedCurrencyPairIDRef_ != ""){
             return boost::shared_static_cast<QuotedCurrencyPair>(IDManager::instance().getID(quotedCurrencyPairIDRef_));
        }else{
             return this->quotedCurrencyPair_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<QuotedCurrencyPair>();
   }
}
boost::shared_ptr<PositiveDecimal> ExchangeRate::getRate()
{
   if(!this->rateIsNull_){
        if(rateIDRef_ != ""){
             return boost::shared_static_cast<PositiveDecimal>(IDManager::instance().getID(rateIDRef_));
        }else{
             return this->rate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PositiveDecimal>();
   }
}
boost::shared_ptr<PositiveDecimal> ExchangeRate::getSpotRate()
{
   if(!this->spotRateIsNull_){
        if(spotRateIDRef_ != ""){
             return boost::shared_static_cast<PositiveDecimal>(IDManager::instance().getID(spotRateIDRef_));
        }else{
             return this->spotRate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PositiveDecimal>();
   }
}
boost::shared_ptr<XsdTypeDecimal> ExchangeRate::getForwardPoints()
{
   if(!this->forwardPointsIsNull_){
        if(forwardPointsIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(forwardPointsIDRef_));
        }else{
             return this->forwardPoints_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
boost::shared_ptr<PointValue> ExchangeRate::getPointValue()
{
   if(!this->pointValueIsNull_){
        if(pointValueIDRef_ != ""){
             return boost::shared_static_cast<PointValue>(IDManager::instance().getID(pointValueIDRef_));
        }else{
             return this->pointValue_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PointValue>();
   }
}
std::vector<boost::shared_ptr<CrossRate>> ExchangeRate::getCrossRate()
{
   if(!this->crossRateIsNull_){
        if(crossRateIDRef_ != ""){
             return this->crossRate_;
        }else{
             return this->crossRate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<CrossRate>>();
   }
}
}

